CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+    of all Topology Director nodes and all the nodes belongining to the relevant Failover Coordinator.  
+    Full topology will be retrieved later (for most nodes from their Failover Coordinator) in order to minimize load on the Topology Director in very large 
+    clusters.
+  
+With the Failover Coordinators for topology & primary/replica state propagation:
+
+**FLOTILLA.FAILOVER_STATUS**: Sent periodically from every Failover Coordinator
+
+    Includes the last known cluster epoch.  
+    Lists all nodes owned by the Failover Coordinator:
+    - The last configuration epoch they acknowledged   
+    - Their primary/replica status
+    - Potentially (but not needed for Flotilla operations) - metadata like:
+        - Replication offset
+        - Last heard timestamp
+     
+    The response includes:
+    - The last failover status received from all other Failover Coordinators
+    - If epoch is outdated, the current topology
+
+
+### The Failover Coordinator
+
+The Failover Coordinator allows individual nodes to achieve consensus regarding Primary / Replica roles within the shard
+they belong to and also propagates topology state changes to all the nodes it is responsible for.
+
+It periodically sends a snapshot of the primary/replica status of all the nodes it owns to the Topology Director, and
+receives topology updates as the response.
+
+#### Interfaces
+
+The Failover Coordinator does not expose an external interface, it's entirely internal to Flotilla and beyond troubleshooting there's no
+need for the administrator to interact directly with it.
+
+- With the Topology Director for topology & primary/replica state propagation
+- With the nodes for heartbeat, leadership change requests & topology propagation.
+
+### The Node
+
+Each node is responsible to periodically send heartbeat information to the Failover Coordinator which owns it. The response it receives
+contains the updated topology & last heartbeat times of the other nodes in the same shard (the group of nodes
+owning the same slots). The node is responsible for promoting itself by writing the updated status to the Failover Coordinator if it
+detects that its current primary has not sent a heartbeat for a long enough time.
+
+#### Interfaces
+
+- With the Failover Coordinator for topology updates and setting primary status
+- With the Topology Director for registering into the cluster
+
+**FLOTILLA.HEARTBEAT**: Detailed in depth under [Failover](#failover)
+
+## Flows
+
+### Membership and role management
+
+All membership management is performed against the Topology Director. The change is then propagated to all other nodes via the Failover Coordinators.
+The Topology Director is responsible for preventing *concurrent* changes to the same node / shard.
+Concurrent here means changes which have not yet been propagated to the entity in question and acknowledged by it via the
+known epoch field in the node(s) heartbeat.
```

Comment:
As a get out of jail, sure -- but it would come with a big warning sign that doing it when the node is still alive might cause data loss.

Reply:
Yeah, I suppose we need an entire separate section for. "Everything is broken, what do!". Which, for better or worse, Redis cluster actually supports really well since it's base APIs are so sharp. It's very easy to unwedge something that ends up in a crazy state.

Reply:
Well... one of the main design goals here is to have a lot less 'wedging' :)
But yes, we will definitely need an extensive set of troubleshooting tools + a tutorial